discord_bot/discord_bot.py

```diff
@@ -1,60 +1,88 @@
+        # actions
+        self.event_loop.run_until_complete(self.gateway_handler())
 
     @staticmethod
     def load_config():
+        """
+        Loads the configurations
+        :return: dict - the configurations
+        """
         with open(CONFIG_FILE_PATH) as f:
             return json.load(f)
 
     @staticmethod
     def get_gateway():
+        """
+        Caches a gateway value and retrieves a new URL
+        :return: gateway URL
+        """
         r = requests.get(DISCORD_GATEWAY_ENDPOINT)
         return r.json()['url']
 
-    async def handshake(self, websocket):
-        identity_obj = {
-            "token": self.config["bot_token"],
-            "properties": {
-                "$os": "windows",
-                "$browser": "disco",
-                "$device": "disco"
-            },
-            "compress": False,
-            "large_threshold": 250,
-            "shard": [1, 10]
-        }
-        await websocket.send(json.dumps({"op": 2, "d": identity_obj}))
+    async def send_json(self, payload):
+        await self.websocket.send(json.dumps(payload))
 
-    async def gateway_handler(self):
+    async def handshake(self, message):
+        """
+        When a websocket connection is opened, Hello payload is received. Then, an Identify payload is sent as
+        part of the handshake to authorize this client.
 
+        :param message: dict - hello payload
+        """
+        self.heartbeat_interval = message["d"]["heartbeat_interval"]
+        await self.heartbeat()
+
+        handshake_identity = self.config["handshake_identity"]
+        await self.send_json({"op": Opcodes.IDENTIFY, "d": handshake_identity})
+
+    async def heartbeat(self):
+        """
+        Sends a heartbeat payload every heartbeat interval.
+        """
+        # await asyncio.sleep(self.heartbeat_interval / 1000.0)
+        await asyncio.sleep(5)
+        await self.send_json({"op": Opcodes.HEARTBEAT, "d": self.last_seq})
+        self.event_loop.create_task(self.heartbeat())
+
+    async def gateway_handler(self):
+        """
+        Creates the websocket, receives messages and acts on them.
+        """
         async with websockets.connect("{}/?v={}&encoding={}".format(self.gateway_ws_url,
                                                                     self.config["gateway_api_version"],
                                                                     self.config["gateway_encoding"])) as websocket:
+            self.websocket = websocket
             while True:
-                message = await websocket.recv()
-                print(message)
+                message = await self.websocket.recv()
                 message = json.loads(message)
-                if message["op"] == 10:
-                    await self.handshake(websocket)
-
-
-# main entry point
-if __name__ == "__main__":
-    m = DiscordBot()
+                self.last_seq = message["s"]
```

Comment:
Notice that `HEARTBEAT_ACK` returns last seq `None`, however we should only send `None` if we haven't received a sequence yet according to the spec, so this is a bug.

Reply:
Don't forget to fix this!